lib/iris/experimental/stratify.py

```diff
@@ -0,0 +1,179 @@
+        of ``phenom``.
+
+    interpolator : callable or None
+        The interpolator to use when computing the interpolation. The function
+        will be passed the following positional arguments::
+
+            (target levels, level data, phenom data, interp axis)
+
+        If the interpolator is None, :func:`stratify.interpolate` will be used
+        with linear interpolation and NaN extrapolation.
+
+        An example of constructing an alternative interpolation scheme:
+
+            from functools import partial
+            interpolator = partial(stratify.interpolate,
+                                   interpolation=stratify.INTERPOLATE_NEAREST,
+                                   extrapolation=stratify.EXTRAPOLATE_LINEAR)
+
+    """
+    # Identify the z-coordinate within the phenom cube
+    if isinstance(interp_direction, Coord):
+        [interp_dim] = phenom.coord_dims(interp_direction)
+    else:
+        interp_dim = interp_direction
+
+    if isinstance(level_data, Coord):
+        source_data = level_data.points
+    else:
+        source_data = level_data.data
+
+    # phenom and level_data must be broadcastable.
+    phenom_data, source_data = np.broadcast_arrays(phenom.data,
+                                                   source_data)
+
+    target_levels = np.array(target_levels)
+    target_phenom_dim = interp_dim
+    if target_levels.ndim != 1:
+        # target levels must be broadcastable to phenom in everything
+        # but the interpolation axis (if not, raise)
+        t_dim_delta = phenom_data.ndim - target_levels.ndim
+        tl_axis = interp_dim - t_dim_delta
+        data_shape = list(phenom_data.shape)
+        data_shape.pop(interp_dim)
+        target_shape = list(target_levels.shape)
+        target_shape.pop(tl_axis)
+        np.broadcast_arrays(np.empty(data_shape), np.empty(target_shape))
+        target_phenom_dim = list(range(phenom_data.ndim))[t_dim_delta:]
+
+    if interpolator is None:
+        interpolator = partial(stratify.interpolate,
+                               interpolation='linear', extrapolation='nan')
+
+    interpolated_phenom = interpolator(target_levels,
+                                       source_data, phenom_data,
+                                       axis=interp_dim)
+
+    # Create a result Cube with the correct shape and metadata.
+    result = Cube(interpolated_phenom, **phenom.copy().metadata._asdict())
+    # Copy across non-Z coordinates.
+    _copy_coords_without_dim(phenom, result, interp_dim)
+
+    coord_kwargs = dict(standard_name=level_data.standard_name,
+                        long_name=level_data.long_name,
+                        var_name=level_data.var_name,
+                        units=level_data.units,
+                        attributes=level_data.attributes)
+    # Add our new interp-dim coordinate.
+    try:
+        result.add_dim_coord(DimCoord(target_levels, **coord_kwargs),
+                             interp_dim)
```

Comment:
@pelson For consistency, it might be best to use `target_phenom_dim` here rather than `interp_dim` ...